Editorial: Refactor handling of numeric-like styles, support differing hours/minutes vs minutes/seconds separators

ben-allen opened the pull request:
- Refactors handling of "numeric"-like styles
- Add additional slots to handle 2-digit hours in locales that use them in all cases
- Add additional slots to handle differing hours/minutes separators in some locales

Fixes https://github.com/tc39/proposal-intl-duration-format/issues/161

ryzokuken commented on spec.emu:
```diff
@@ -234,6 +236,8 @@ contributors: Ujjwal Sharma, Younies Mahmoud
             1. Throw a *RangeError* exception.
           1. If _unit_ is *"minutes"* or *"seconds"*, then
             1. Set _style_ to *"2-digit"*.
+        1. If _unit_ is *"hours"* and _durationFormat_.[[TwoDigitHours]] is *true*, then
+          1. Set _style_ to *"2-digit"*.
```
nit: it could be merged with the condition above
```suggestion
          1. If _unit_ is *"minutes"* or *"seconds"* or _unit_ is *"hours"* and _durationFormat_.[[TwoDigitHours]] is *true*, then
          1. Set _style_ to *"2-digit"*.
```

ben-allen replied:
My thought is to leave this one as-is, because stringing together four conditionals on one line will likely be annoying for implementers

ryzokuken commented on spec.emu:
```diff
@@ -495,18 +697,27 @@ contributors: Ujjwal Sharma, Younies Mahmoud
         1. Set _durationFormat_.[[Locale]] to _locale_.
+        1. Set _durationFormat_.[[DataLocale]] to _r_.[[dataLocale]].
+        1. Let _dataLocale_ be _durationFormat_.[[DataLocale]].
+        1. Let _dataLocaleData_ be _durationFormat_.[[LocaleData]].[[&lt;_dataLocale_&gt;]].
```
This doesn't look right?

ryzokuken replied:
Also you no longer need `dataLocaleData` IIUC, you can directly extract the value of `digitalFormat` and set the values in the instance from that.

ben-allen replied:
going to leave this one for now -- I'm working on a PR to update DurationFormat to reflect changes made in https://github.com/tc39/ecma402/pull/849

ryzokuken commented on spec.emu:
```diff
@@ -268,93 +272,230 @@ contributors: Ujjwal Sharma, Younies Mahmoud
+
+    <emu-clause id="sec-listformatparts" type="abstract operation">
+      <h1>ListFormatParts (
+          _nfResult_: a List of Records,
```
I'd be in favor of generalizing this AO by renaming things from `nfResultFoo` which is specific to our use case (we only list format the results of NF) to something more general.

ryzokuken commented on spec.emu:
```diff
@@ -363,27 +504,88 @@ contributors: Ujjwal Sharma, Younies Mahmoud
+          1. If _unit_ is *"seconds"*, *"milliseconds"*, or *"microseconds"*, then
+            1. If NextUnitFractional(_durationFormat_, _duration_, _unit_) is *true*, then
+              1. Set _value_ to _value_ + AddFractionalDigits(_durationFormat_, _duration_).
+              1. If _durationFormat_.[[FractionalDigits]] is *undefined*, then
+                1. Let _maximumFractionDigits_ be *9*<sub>𝔽</sub>.
+                1. Let _minimumFractionDigits_ be *+0*<sub>𝔽</sub>.
+              1. Else,
+                1. Let _maximumFractionDigits_ be _durationFormat_.[[FractionalDigits]].
+                1. Let _minimumFractionDigits_ be _durationFormat_.[[FractionalDigits]].
+              1. Perform ! CreateDataPropertyOrThrow(_nfOpts_, *"maximumFractionDigits"*, _maximumFractionDigits_).
+              1. Perform ! CreateDataPropertyOrThrow(_nfOpts_, *"minimumFractionDigits"*, _minimumFractionDigits_).
+              1. Perform ! CreateDataPropertyOrThrow(_nfOpts_, *"roundingMode"*, *"trunc"*).
```
Didn't you move this into an AO? Idk how it ended up back in.

ryzokuken commented on spec.emu:
```diff
@@ -268,93 +272,230 @@ contributors: Ujjwal Sharma, Younies Mahmoud
+        1. Let _secondsValue_ be _duration_.[[Seconds]].
+        1. Let _fractionalDigits_ be AddFractionalDigits(_durationFormat_, _duration).
+        1. If _fractionalDigits_ is not 0, then
+          1. If _durationFormat_.[[FractionalDigits]] is *undefined*, then
```
Replace with `SetFractionalDigitOptions`

ben-allen replied:
See above

ryzokuken commented on spec.emu:
```diff
@@ -363,27 +504,88 @@ contributors: Ujjwal Sharma, Younies Mahmoud
+          1. If _unit_ is *"seconds"*, *"milliseconds"*, or *"microseconds"*, then
+            1. If NextUnitFractional(_durationFormat_, _duration_, _unit_) is *true*, then
+              1. Set _value_ to _value_ + AddFractionalDigits(_durationFormat_, _duration_).
+              1. If _durationFormat_.[[FractionalDigits]] is *undefined*, then
```
Replace with `SetFractionalDigitOptions`

ryzokuken reviewed (changes requested):
Massive improvement, kudos! A bunch of comments inline to make this PR perfect :D

ben-allen commented:
Supersedes #186

sffc commented:
TG2 discussion: https://github.com/tc39/ecma402/blob/main/meetings/notes-2024-02-22.md#durationformat-update

FrankYFTang commented on spec.emu:
```diff
@@ -189,6 +190,7 @@ contributors: Ujjwal Sharma, Younies Mahmoud
     <emu-clause id="sec-getdurationunitoptions" type="abstract operation">
       <h1>
         GetDurationUnitOptions (
+          _durationFormat_: a DurationFormat Object,
```
Why do we need to pass in a  DurationFormat Object here?  I see you only use _durationFormat_.[[TwoDigitHours]]  inside this AO. Why not just pass in that value only ?

FrankYFTang commented on spec.emu:
```diff
@@ -268,93 +272,231 @@ contributors: Ujjwal Sharma, Younies Mahmoud
-        PartitionDurationFormatPattern (
-          _durationFormat_: a DurationFormat,
+        SetFractionalDigitOptions (
+        _durationFormat_: a DurationFormat Object,
```
The only thing you use internally is _durationFormat_.[[FractionalDigits]], why not just pass in that value and avoid passing in _durationFormat_ ?

FrankYFTang commented on spec.emu:
```diff
@@ -268,93 +272,231 @@ contributors: Ujjwal Sharma, Younies Mahmoud
+    <emu-clause id="sec-formathours" type="abstract operation">
+      <h1>
+        FormatHours (
+          _durationFormat_: a DurationFormat Object,
```
how about we only pass in  _durationFormat_.[[Locale]] and _durationFormat_.[[HoursStyle]] but not passing in _durationFormat_?

ben-allen commented:
Incorporated feedback, made a few additional changes:
* renamed `isFirstUnitDisplayed` and `isSignDisplayed` aliases to `signDisplayed` throughout
* Renamed `FormatHours`, `FormatMinutes`, `FormatSeconds` to `FormatNumericHours`, `FormatNumericMinutes`, `FormatNumericSeconds` to clarify usage
* Removed `SetFractionalDigitOptions`

Justification for removing `SetFractionalDigitOptions`: Talked it over with @ptomato, who observed that it could be confusing to add `nfOpts` properties in multiple AOs due to how setting properties can trigger setters in user code. Options as I see it:

1. Leave as in this version of PR, with the steps for setting fractional digit options repeated in two places.
2. Restore `SetFractionalDigitOptions`, since we're passing `*null*` to `OrdinaryObjectCreate` when creating `nfOpts` objects and so, if I understand it correctly, the poisoned object problem won't actually occur. However: creating properties on objects across multiple AOs is not currently something that happens in either 402 or Temporal
3. Rewrite `SetFractionalDigitOptions` to return a Record which is then used by the caller to create `nfOpts` properties. Note: this doesn't actually save that many repeated steps.

requesting re-review! @ryzokuken @FrankYFTang

ben-allen commented:
Requesting re-review! @FrankYFTang @ryzokuken

ryzokuken commented on spec.emu:
```diff
@@ -363,27 +477,89 @@ contributors: Ujjwal Sharma, Younies Mahmoud
+                1. Append the Record { [[Type]]: _part_.[[Type]], [[Value]]: _part_[[Value]], [[Unit]]: _numberFormatUnit_ } to _list_.
+              1. Append _list_ to _result_.
+        1. Return ListFormatParts(_durationFormat_, _result_).
+      </emu-alg>
```
Great! PDFP (if I can call it that) has become so much better!

ryzokuken commented on spec.emu:
```diff
@@ -495,18 +671,27 @@ contributors: Ujjwal Sharma, Younies Mahmoud
+        1. Set _durationFormat_.[[DataLocale]] to _r_.[[dataLocale]].
+        1. Let _dataLocale_ be _durationFormat_.[[DataLocale]].
+        1. Let _dataLocaleData_ be _durationFormat_.[[LocaleData]].[[&lt;_dataLocale_&gt;]].
+        1. Let _digitalFormat_ be _dataLocaleData_.[[DigitalFormat]].
```
Basically this is the value you actually need, the few lines right before this one can be collapsed into this.

ben-allen replied:
I've thought pretty hard about this one. Collapsing these lines together doesn't match rest-of-402 style, as I understand it. See:

* steps 18 and 19 of [`Intl.DisplayNames` constructor](https://tc39.es/ecma402/#sec-Intl.DisplayNames):]

        18. Let _resolvedLocaleData_ be _r_.[[LocaleData]].
        19. Let _types_ be _resolvedLocaleData_.[[types]].

with _resolvedLocaleData_ never used after the [[types]] access.

* Steps 14 and 15 of [`Intl.ListFormat` constructor](https://tc39.es/ecma402/#sec-Intl.ListFormat), _resolvedLocaleData_ likewise not used after _resolvedLocaleData_.[[<type>]] access:

        14. Let _resolvedLocaleData_ be _r_.[[LocaleData]].
        15. Let _dataLocaleTypes_ be _resolvedLocaleData_.[[&lt;_type_&gt;]].


* First two steps of [`GetNumberFormatPattern` AO](https://tc39.es/ecma402/#sec-getnumberformatpattern) in `Intl.NumberFormat`

        1. Let _resolvedLocaleData_ be _numberFormat_.[[LocaleData]].
        2. Let _patterns_ be _resolvedLocaleData_.[[patterns]].
        
 (among others)

Additionally, the pattern "Let _alias_ be _obj_.[[X]].[[Y]].[[Z]]" is very rare in 402 algorithms; it shows up in one place in `Intl.DateTimeFormat` and nowhere else (https://tc39.es/ecma402/#sec-date-time-style-format)

Four-deep accesses (as would happen here if these were collapsed together) occur not at all in rest-of-402

ryzokuken commented on spec.emu:
```diff
@@ -495,18 +671,27 @@ contributors: Ujjwal Sharma, Younies Mahmoud
+        1. Let _dataLocale_ be _durationFormat_.[[DataLocale]].
+        1. Let _dataLocaleData_ be _durationFormat_.[[LocaleData]].[[&lt;_dataLocale_&gt;]].
+        1. Let _digitalFormat_ be _dataLocaleData_.[[DigitalFormat]].
+        1. Let _twoDigitHours_ be _digitalFormat_.[[TwoDigitHours]].
+        1. Set _durationFormat_.[[TwoDigitHours]] to _twoDigitHours_.
+        1. Let _hoursMinutesSeparator_ be _digitalFormat_.[[HoursMinutesSeparator]].
+        1. Set _durationFormat_.[[HoursMinutesSeparator]] to _hoursMinutesSeparator_.
+        1. Let _minutesSecondsSeparator_ be _digitalFormat_.[[MinutesSecondsSeparator]].
+        1. Set _durationFormat_.[[MinutesSecondsSeparator]] to _minutesSecondsSeparator_.
```
No need to set three variables, all of these lines can be collapsed.

ryzokuken reviewed (changes requested):
Overall LGTM, the improvements to the spec are quite evident by the amount of time it took me to review it in entirety this time... Either that or there's something with my coffee :P

Anyway, I have made a flurry of comments that should all further clean up the text. All of this is very much editorial and fills me with hope that by the end of this exercise we'd have a pretty looking and readable spec.

ben-allen commented on spec.emu:
```diff
@@ -268,93 +271,204 @@ contributors: Ujjwal Sharma, Younies Mahmoud
+        1. Let _nf_ be ! Construct(%NumberFormat%, &laquo; _durationFormat_.[[Locale]] &raquo;, _nfOpts_).
+        1. Let _hoursParts_ be ! PartitionNumberPattern(_nf_, _hoursValue_).
+        1. For each Record { [[Type]], [[Value]] } _part_ of _hoursParts_, do
+          1. Return the Record { [[Type]]: _part_.[[Type]], [[Value]]: _part_.[[Value]], [[Unit]]: *"hour"* }.
```
oof, can't believe I left that in

ben-allen commented on spec.emu:
```diff
@@ -268,93 +271,204 @@ contributors: Ujjwal Sharma, Younies Mahmoud
+          1. Let _hoursFormatted_ be *true*.
+        1. Else,
+          1. Let _hoursFormatted_ be *false*.
+        1. If _secondsValue_ is not 0 or _secondsDisplay_ is not *"auto"*, then
+          1. Let _secondsFormatted_ be *true*.
+        1. Else,
+          1. Let _secondsFormatted_ be *false*.
+        1. If _hoursFormatted_ is *true* and _secondsFormatted_ is *true*, then
```
In this case it's necessary to declare these aliases out of order, since `_minutesFormatted_` depends on `_hoursFormatted_` and `_secondsFormatted_`

ben-allen commented on spec.emu:
```diff
@@ -268,93 +271,204 @@ contributors: Ujjwal Sharma, Younies Mahmoud
+          1. Append the Record { [[Type]]: _part_.[[Type]], [[Value]]: _part_.[[Value]], [[Unit]]: *"second"* } to _result_.
+        1. Return _result_.
+        </emu-alg>
+    </emu-clause>
```
`FormatNumericHours` and `FormatNumericMinutes` are combinable, but `FormatNumericSeconds` is pretty different from both due to the extra properties it has to set on `_nfOpts_`. It could save some repeated code to combine the first two into `FormatNumericHoursMinutes`, though -- I'll give it a try.

ben-allen commented on spec.emu:
```diff
@@ -363,27 +477,89 @@ contributors: Ujjwal Sharma, Younies Mahmoud
+              1. If NextUnitFractional(_durationFormat_, _unit_) is *true*, then
+                1. Set _value_ to _value_ + AddFractionalDigits(_durationFormat_, _duration_).
+                1. If _durationFormat_.[[FractionalDigits]] is *undefined*, then
+                  1. Let _maximumFractionDigits_ be *9*<sub>𝔽</sub>.
+                  1. Let _minimumFractionDigits_ be *+0*<sub>𝔽</sub>.
```
I don't see where those values could get set -- isn't `[[FractionDigits]] assigned a value returned from `GetNumberOption` with `*undefined*` as the default?

ben-allen commented on spec.emu:
```diff
@@ -234,6 +235,8 @@ contributors: Ujjwal Sharma, Younies Mahmoud
             1. Throw a *RangeError* exception.
           1. If _unit_ is *"minutes"* or *"seconds"*, then
             1. Set _style_ to *"2-digit"*.
+        1. If _unit_ is *"hours"* and _twoDigitHours_ is *true*, then
```
We've had the previous lines about forcing minutes and seconds to `"2-digit"` when preceded by a `"numeric"` or `"2-digit"` style in the spec for quite some time, and I view this as analogous. I'm working on the assumption -- which may be incorrect -- that in the locales that always use a two-digit representation of hours, displaying a one-digit hour is analogous to using the wrong separator.

ben-allen commented:
I've pushed a major update reflecting feedback from @ryzokuken and incorporating some additional bugfixes. 

I'd like to draw attention to the last commit: c6025ef4321f582beb5fcc68d2bc5e8a66d9ce02. In this one I removed an alias that was used to ensure that the fractional digit-related options on the `_nfOpts_` used by the `Intl.NumberFormat` in `FormatNumericSeconds` weren't added if there weren't actually any nonzero subsecond values. 

 I removed this one because the previous versions of the spec *do* in fact add those properties whenever there are any subsecond units styled as `'fractional'` (or in earlier revisions, `"numeric"`) exist, regardless of whether or not any of them hold a nonzero value.

This strikes me as the wrong behavior -- since subsecond units *must* be styled as `"fractional"` whenever hours, minutes, or seconds are styled as `"numeric"`, the fractional digit options for `NumberFormat` are always added when numeric styles are used. 

Any changes to this behaviour would be observable.

ben-allen commented:
New revision ready for review! Much more solid thanks to @ryzokuken's feedback, but could still use more eyes. @gibson042 @FrankYFTang @sffc @anba

sffc commented:
TG2 notes: https://github.com/tc39/ecma402/blob/main/meetings/notes-2024-04-25.md#durationformat-update

sffc reviewed (approved):
I reviewed this in enough detail to say that I believe it covers the edge cases we had identified and overall makes the spec much easier to reason about and implement. Please merge :)
